temporal: Update STDS metadata from DB

ninsbl opened the pull request:
After digging a bit I figured out that the bottleneck and reason for the very slow process for updating STDS metadata (see: #3136) is the _get_registered_maps_as_objects_ function.

Creating a map object takes 1/10 og 1/20 of a second, but with 100k maps that sums up to a quite a bit of time...

This PR is a work around for updating STDS metadata, that avoids using the slow function. But ideally, the function itself (_get_registered_maps_as_objects_) should be improved so it scales to larger STDS... I assume it is initilizing of DB interface that is most time consuming, but I did not investigate more deeply...

ninsbl commented:
@wenzeslaus and @veroandreo , as you may have seen from the recent commits, I have not yet testet fully locally. When / if you look into this, please check for now mainly the general concept. I will update the PR once I have a full local test done...

ninsbl commented:
OK, after some local testing it seems I managed to iron out the issues, so it should be ready to review.

@wenzeslaus I wonder if you have any suggestions / opinions on where to put the new helper functions? Maybe better in _datetime_math_?

tmszi commented:
I tested it with 4000 registered maps, where the original code for registering a new map took about 13 seconds on my computer, with your changes it took 0.972 seconds. I haven't looked at the implementation details of your code yet.

ninsbl commented:
> I tested it with 4000 registered maps, where the original code for registering a new map took about 13 seconds on my computer, with your changes it took 0.972 seconds. I haven't looked at the implementation details of your code yet.

Thanks for testing @tmszi ! Do you have any suggestions where to put the new __get_map_time_tuple_, __get_row_time_tuple_ and __is_after_ functions? Are names and placement OK? My experience with developing at the library level is still somewhat limited....

echoix commented:
I think that this comment https://github.com/OSGeo/grass/pull/3359#discussion_r1451817994 meant that you wanted input on the `get_registered_maps_as_objects` function. 

Since the code from the function `get_registered_maps_as_objects` wasn't changed in the PR, I can't comment directly on it. 
What does checking for db_version to in here? It doesn't seem to be used anywhere, and textual usages of db_version exist only in C code.
https://github.com/OSGeo/grass/blob/0a122482c540f7aeb4176d44660967685c1a601e/python/grass/temporal/abstract_space_time_dataset.py#L1541-L1553
If it really is needed, then if the data structure allows it, would it be possible to skip the python for loop and use a more direct lookup? or even a comprehension of some kind, since they have a different path than for loops internally.

echoix commented:
I'm currently running gunittest on the python/grass/temporal/ folder through a python debugger to find a test that could call that function and could see what happens. For now I'm at my third try and still got jumped so much all around the code so many times, and I'm still only inside this line: https://github.com/OSGeo/grass/blob/0a122482c540f7aeb4176d44660967685c1a601e/python/grass/temporal/abstract_space_time_dataset.py#L1554-L1557

ninsbl commented:
Here is how to reproduce / test:
```
# Set a small computational region
g.region -p n=15 s=0 e=15 w=0

# Create an empty STRDS
t.create output=test_strds type=strds temporaltype=absolute semantictype=mean title=None description=None

# Create many raster maps
for i in $(seq 1 5000); do r.mapcalc expression="map_a${i}=5000+$i" --o --v; done

# Write til long list of raster maps to file
g.list type=raster pattern=map_* > reg.txt

# Register the maps
t.register input=test_strds start="2024-01-10" increment="1 day" file=reg.txt
```

Then the TGIS SQLite DB should be in $GISDBASE/$LOCATION/$MAPSET/tgis/sqlite.db (if you did not chose something different)....

I will see if I can upload a TGIS DB, but that doees not allow full testing without access to the underlying maps...

ninsbl commented:
> I think that this comment [#3359 (comment)](https://github.com/OSGeo/grass/pull/3359#discussion_r1451817994) meant that you wanted input on the `get_registered_maps_as_objects` function.

Actually, my main questions are:

1) Are names of the introduced `_get_map_time_tuple`, `_get_row_time_tuple` and `_is_after` functions OK (not sure about when to use `_` in function names?
2) Is it OK to have them where I put them? They could probably go into `datetime_math` instead?

> What does checking for db_version to in here? It doesn't seem to be used anywhere, and textual usages of db_version exist only in C code.
> If it really is needed, then if the data structure allows it, would it be possible to skip the python for loop and use a more direct lookup? or even a comprehension of some kind, since they have a different path than for loops internally.

TGIS DB < 3 does not have `semantic_labels` as a meta data element. So those DBs from different versions have different metadata columns. I changed the check however in #3359

echoix commented:
> Here is how to reproduce / test:
> 
> ```
> 
> # Set a small computational region
> 
> g.region -p n=15 s=0 e=15 w=0
> 
> 
> 
> # Create an empty STRDS
> 
> t.create output=test_strds type=strds temporaltype=absolute semantictype=mean title=None description=None
> 
> 
> 
> # Create many raster maps
> 
> for i in $(seq 1 5000); do r.mapcalc expression="map_a${i}=5000+$i" --o --v; done
> 
> 
> 
> # Write til long list of raster maps to file
> 
> g.list type=raster pattern=map_* > reg.txt
> 
> 
> 
> # Register the maps
> 
> t.register input=test_strds start="2024-01-10" increment="1 day" file=reg.txt
> 
> ```
> 
> 
> 
> Then the TGIS SQLite DB should be in $GISDBASE/$LOCATION/$MAPSET/tgis/sqlite.db (if you did not chose something different)....
> 
> 
> 
> I will see if I can upload a TGIS DB, but that doees not allow full testing without access to the underlying maps...

That's enough to have something to work with ;)

ninsbl commented:
BTW, relevant tests are:
https://github.com/OSGeo/grass/blob/main/python/grass/temporal/testsuite/test_register_function.py
and
https://github.com/OSGeo/grass/tree/main/temporal/t.register/testsuite

echoix commented:
Just by reading like this, I don't see something that would really have any impact. I'd need to do some profiling to know what's going on and where to look and find the bottlenecks. I haven't done some Python profiling without PyCharm Professional (I really liked it) that I had access as a student while still at university. Now I'd have to find a way of getting the same info with another tool for that.

It won't be for tonight though.

ninsbl commented:
OK, now I did some profiling with builtin Python profiling tools:

```
from cProfile import Profile
from pstats import SortKey, Stats
import grass.temporal as tgis
tgis.init()
dbif = tgis.SQLDatabaseInterfaceConnection()
dbif.connect()
sp = tgis.open_old_stds('test_strds', "strds", dbif)
with Profile() as profile:
    print(f"{ sp.get_registered_maps_as_objects() = }")
    (Stats(profile).strip_dirs().sort_stats(SortKey.TIME).print_stats())
```
This is the most relevant result (with a different dataset than the example above):

```
# (...)
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1111   20.619    0.019   20.619    0.019 {method 'execute' of 'sqlite3.Cursor' objects}
     2220    5.075    0.002    5.075    0.002 {built-in method posix.read}
      444    0.755    0.002    0.755    0.002 {built-in method _sqlite3.connect}
      223    0.161    0.001    0.166    0.001 {method 'fetchall' of 'sqlite3.Cursor' objects}
 ```

So, it seems there is quite som DB querying going on, although all relevant information is already fetched.

Building / initializing a single RasterDataset:

```
with Profile() as profile:
    print(f"{ sp.get_new_map_instance('id@id') = }")
    (Stats(profile).strip_dirs().sort_stats(SortKey.TIME).print_stats())
```
involves already 2 connections, 5 execute calls to SQLite and 10 reads. Given that all information is available in a row of the relevant DB view that should not be necessary... 

```
 sp.get_new_map_instance('id@id') = <grass.temporal.space_time_datasets.RasterDataset object at 0x7fe317210ac0>
         524 function calls in 0.124 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        5    0.061    0.012    0.061    0.012 {method 'execute' of 'sqlite3.Cursor' objects}
       10    0.053    0.005    0.053    0.005 {built-in method posix.read}
        2    0.004    0.002    0.004    0.002 {built-in method _sqlite3.connect}
        4    0.001    0.000    0.001    0.000 {built-in method posix.stat}
        1    0.001    0.001    0.001    0.001 {method 'fetchall' of 'sqlite3.Cursor' objects}
```

Next, I tested the following sub-function calls within _get_registered_maps_as_objects_, and those do not seem to communicate with SQLite and they do not seem to consume a noticable amount of time (0.000). So,  the _get_new_map_instance_ function, seems to be what is causing the parent function to be slow... Or more specific the initialization of the `RasterDataset`...

echoix commented:
Isn't the goal of `get_registered_maps_as_objects ` to create such Dataset object for each of the elements (rows or time steps, whatever it is called in that domain).
I'm not too surprised seeing Rasterdataset, it appeared a bit when stepping into the code.


And if we take it a little bit higher level, what does a profiling report indicate when doing a more realistic work, including multiple high level calls/operations? Like if we find out that a slow function is called another separate time when it could be simplified? Like if `get_registered_maps_as_objects ` gets called too many times that we weren't aware of.

echoix commented:
A bit of reading and other future ideas/possible avenues to explore some day.

At some places using context managers (that can close connections) might help a bit with the dbif always getting created and closed inside a same function
https://docs.python.org/3/library/contextlib.html#contextlib.closing

Like here 
https://github.com/ninsbl/grass/blob/110ff36148347bcb81aeb0f8ca506ca21a6768f4/python/grass/temporal/core.py#L678-L696
Followed in the same function with

https://github.com/ninsbl/grass/blob/110ff36148347bcb81aeb0f8ca506ca21a6768f4/python/grass/temporal/core.py#L731-L739
And other function calls inside the same function that may (by only reading) also create and close connections. 


The context manager pattern looks very similar to what is really commonly used in dotnet, especially for ressources that needs to be released after, whatever happens.




That post was quite informative
https://stackoverflow.com/questions/53388451/how-to-speed-up-python-instance-initialization-for-millions-of-objects

It pointed me to the existence of `__slots__`, 
https://docs.python.org/3/reference/datamodel.html#slots
, similar to `__dict__` and `__all__` (that we can see sometimes in `__init__.py`), that allows to explicitly set what data members (like fields and properties) are allowed to exist, and disables the normal usage of `__dict__` for this. It takes less memory, and less work must be made to retrieve. I'm keeping this in mind if I see a potential use case.

There's also this, dataclasses
https://docs.python.org/3/library/dataclasses.html, that help out creating repetitive  classes for data, and having common nice-to-have functions already implemented. Again, I don't know if we have a use case yet for that. It shouldn't be faster though, but less code for us.


Is this long to run? Does every RasterDataset need to have all the types created?

https://github.com/ninsbl/grass/blob/110ff36148347bcb81aeb0f8ca506ca21a6768f4/python/grass/temporal/space_time_datasets.py#L243-L250

veroandreo commented:
I checked-out this pr with `gh pr checkout 3350`, then configured and compiled; and then executed the commands you provided. The registering takes around 3 minutes... What am I doing wrong?

ninsbl commented:
> I checked-out this pr with `gh pr checkout 3350`, then configured and compiled; and then executed the commands you provided. The registering takes around 3 minutes... What am I doing wrong?

Hi @veroandreo , I don` t think you do anything wrong (I do not know/use the GH CLI). Registering many maps still takes some time (I guess it can be improved too). But if you register e.g. 5 new maps into _test_strds_, you should notice a difference to before the PR...

Try e.g.:

```
# Create 5 new raster maps
for i in $(seq 1 5); do r.mapcalc expression="map_n${i}=$i" --o --v; done

# Write til long list of raster maps to file
g.list type=raster pattern=map_n* > reg.txt

# Register the maps

t.register input=test_strds start="2024-01-01" increment="1 day" file=reg.txt
```

ninsbl commented:
It seems the TGIS DB version check from DB metadata in _metadata.py_ is the main reason for time consuming database reads. Thanks @echoix for helping with digging that up.
I don`t think this check is needed. During intialization of TGIS, it is checked if the DB version does not match the library version, and users get noticed that versions are incompatible... In all other cases set_semantic_labels is only used if TGIS supports i, so, it should be OK to remove that check for every RasterDataset... Or is there a reason to keep this costly check, @landam ?

ninsbl commented:
How do we get this over the finish line?

This is a major limitation for the scalability of many temporal modules / TGIS functionality!

echoix commented on python/grass/temporal/abstract_space_time_dataset.py:
```diff
@@ -1517,45 +1513,30 @@ def get_registered_maps_as_objects(
 
         :return: The ordered map object list,
-                In case nothing found None is returned
+                In case nothing found an empty list is returned
```
```suggestion
                In case nothing is found, an empty list is returned
```

echoix commented on python/grass/temporal/abstract_space_time_dataset.py:
```diff
@@ -1517,45 +1513,30 @@ def get_registered_maps_as_objects(
+            # check keys in first row
+            # note that 'if "bottom" in row' does not work
+            # because row is not a dict but some db backend object
+            has_bt_columns = "bottom" in rows[0].keys() and "top" in rows[0].keys()
```
When I read specifically about the sqlite3 Row class (in Python), a row can be accessed both by index or column name. So technically `rows[0]["bottom"]` could work, and if so something like the following might work (I'm not sure if it's the right test to do with it to have what you want, but an illustration of what it could look like without two other iterations for membership testing, probably only impactful if there are a lot of columns)

```suggestion
            has_bt_columns = rows[0]["bottom"] is not None  and rows[0]["top"] is not None
```

ninsbl replied:
The purpose of this check is to get the data structure, i.e. if the columns "bottom" and "top" are present in the metadata DB view. If "bottom" is present "top" is present too, so checking for "bottom" should be enough...

echoix replied:
Does the code break if there isn't the two columns?

ninsbl replied:
Yes, I think so. I do have some more fixes and will look into `compute_datetime_delta` too, ASAP.

echoix replied:
After refreshing up on the data structures in Python (most of that knowledge dated from the late-Python 2 to Python 3 transition era, so it was a bit far), my initial first impressions on the cost of a membership testing are more relaxed now. If we are talking about dictionaries, `.keys()` returns a dictionary view and thus I suppose calling `in` is like a hash-based lookup, or something that would be similar to what is done for sets. So there wouldn't be a problem even for the original.

https://docs.python.org/3/library/stdtypes.html#dict.keys

https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects

echoix commented on python/grass/temporal/abstract_space_time_dataset.py:
```diff
@@ -2872,7 +2853,7 @@ def update_from_registered_maps(self, dbif=None):
 
         # Count the temporal map types
-        maps = self.get_registered_maps_as_objects(dbif=dbif)
+        maps = self.get_registered_maps(order="start_time", dbif=dbif)
```
Is `get_registered_maps_as_objects` still used anywhere else, or used publicly, or is part of our public API? If that was the last reference, there would be some cleaning to do.

ninsbl replied:
The function is relatively frequently used:
https://github.com/search?q=repo%3AOSGeo%2Fgrass%20get_registered_maps_as_objects(&type=code

https://github.com/OSGeo/grass/pull/3350/commits/125c80e6b197d643c955835f596d212edf399347 should improve the performance quite a bit...

echoix reviewed (changes requested):
I did a 2h+ search and review (on a phone), especially for the granularity file. I tried to follow the code execution by reading, without getting distracted by the implementations or anything else around, just what is inside the function at the time of reading it. So some answers are conditional.

ninsbl commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -221,30 +297,32 @@ def compute_relative_time_granularity(maps):
+                    t = abs(start - start2)
                     delta.append(int(t))
 
     delta.sort()
```
Hmm.... ` {}` creates a dict, Is there a literal for creating an empty set? `set({})` seems less efficient than ` set()` ...

echoix replied:
https://stackoverflow.com/questions/36674083/why-is-it-possible-to-replace-sometimes-set-with

Ya seems like you're right. Well, since it's only for the initialization, let's not worry about it, it remains at the order of micro optimisation.

A side note to take into account when you check the performance when refactoring. Starting with Python 3.10, performance is quite different (better) than 3.9. They worked on a project "Faster Python project", while the 50% at each step wasn't held
https://github.com/markshannon/faster-cpython/blob/master/plan.md
https://github.com/faster-cpython/ideas
There was still big changes under the hood. https://docs.python.org/3.10/whatsnew/3.10.html#optimizations

https://docs.python.org/3.11/whatsnew/3.11.html#optimizations

https://docs.python.org/3.11/whatsnew/3.11.html#whatsnew311-faster-cpython

https://docs.python.org/3.12/whatsnew/3.12.html#optimizations


So, tuning for specific inefficiencies of already old releases that only have security updates for a little while remaining (3.8 is really near the end), doesn't bring as much.

ninsbl commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -351,37 +429,28 @@ def compute_absolute_time_granularity(maps):
+                    datetime_delta.append(compute_datetime_delta(start, start2))
     # Check what changed
     dlist = []
     for d in datetime_delta:
```
`compute_datetime_delta` creates always dictionaries with all time units, so the membership check is not needed actually. 
The first loop over `datetime_delta` is now merged into the initial loop over the maps. However, that loop cannot be exited once _has_seconds_ is _True_...

Also, `datetime_delta` is now a set. Had to convert the datetime_delta dictionaries to tuples, so they can be added to sets (and then back to dicts when used)...

ninsbl commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -351,37 +429,28 @@ def compute_absolute_time_granularity(maps):
-    for map in maps:
-        start, end = map.get_temporal_extent_as_tuple()
+    for i in range(len(maps)):
+        start, end = get_time_tuple(maps[i])
```
No, at least to my knowledge, `compute_relative_time_granularity()` is for _SpaceTimeDatasets_ (and maps therein)  with relative time, and `compute_absolute_time_granularity()` is for _SpaceTimeDatasets_ (and maps therein) with absolute time.

ninsbl commented:
Thanks, @echoix for all your suggestions. I think they improved the code quite a bit! And I learned a lot from that. Let me know if you thonk I addressed your comments sufficiently.

echoix commented:
I'll take a look soon tonight, depending on your time zone you'll see it tomorrow morning. Apart from my review, do you consider it finished and ready?

ninsbl commented:
If all tests fail I would consider it ready. At least it is a noticeable improvement for larger space time raster datasets.

Other use of `get_registered_maps_as_objects` could be improved in a different PR, I guess...

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,143 @@ def compute_absolute_time_granularity(maps):
+        if end:
+            map_datetime_delta = compute_datetime_delta(start, end)
+            has_units = _update_time_unit_dict(has_units, map_datetime_delta)
+            datetime_delta.add(sorted(tuple(map_datetime_delta.items())))
```
Seems like you hit the fact that a set must be hashable objects. https://docs.python.org/3/glossary.html#term-hashable

https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset

My first impression here was to say that `sorted` for a `set` was useless. My second impression was that instead of adding from a `sorted` (that would be a generator), why don't you add all elements at once with datetime_delta.update( _the elements_) https://docs.python.org/3/library/stdtypes.html#frozenset.update. But my third though was that sorted only sorted one single element, since it is a tuple. Does calling `sorted` on a single tuple sort a size 1 sequence or inside the tuple? -> a tuple is immutable and keeps the same order of the iterable. 

https://docs.python.org/3/library/stdtypes.html#tuple

So I assume you wanted to place `sorted()` inside of `tuple`, to sort the items returned by `map_datetime_delta.items()`. (that are themselves tuples, since it is (key, value) for a dictionary), so we end up with a tuple looking like ( ('hour', 8796), ('month',12) , ('second', 0), ('max_days', 366), ('year', 1), ('day', 0), ('minute', 0) )
So it's a kinda weird structure for a set. 


When asking myself why does he do it like that, I understood that you most likely were trying to do the equivalent of converting a dictionary to a `namedtuple`, and populating it with a `**` https://docs.python.org/3/library/collections.html#collections.namedtuple, because you wanted the set to eliminate duplicates and wanted to make sure that the order of .items was the same to allow this. It is good to know that since Python 3.6, `**kwargs` keep their order, and the implementation detail for dicts, is since 3.7 guaranteed to preserve insertion order. 
https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order 

https://docs.python.org/3/library/stdtypes.html#dict.values
(Three sections below, starting with Dictionaries preserve insertion order. )

https://docs.python.org/3/library/collections.html#collections.somenamedtuple._field_defaults (third example underneath that title. 




But with your comment mentioning that the function was a dictionary and you passed through hoops to go and from, I'm pretty sure it isn't the good structure to help. especially since dicts are recreated multiple times s in loops below. 


What about just keeping a list and deduplicate afterwards?

Or maybe if instead of has_unit, it was 

```
unique_units_or_whatever = {
'hours': set(),
'minute': set()
'seconds': set()
}
```

And using dict comprehension

```
unique_units_or_whatever = { k: v.add(map_datetime_delta[k] for k, v in unique_units_or_whatever.items() }
```

That should, if assigning to the same variable that is inside the dict comprehension is allowed, allow to go fetch all previously known fields and update the inner sets (only dict keys need to be hashable, the values don't have to).

ninsbl replied:
Exactly! Using `unique_units_or_whatever` sounds like a very good idea. Will try that!

ninsbl replied:
Looks like this basically works but v.add returns None.

So, this could do the trick:
```
unique_units_or_whatever = {
    'hour': set(),
    'minute': set(),
    'second': set(),
    }

map_datetime_delta = {"hour": 2, "second": 2, "minute": 0}

# Then we can either union two sets and replace the dict from comprehension
unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items() }

# or loop over ordered keys and update in place:
for time_key in ["second", "minute", "hour"]:
    time_val = map_datetime_delta[time_key]
    if time_val > 0:
       unique_units_or_whatever[time_key].add(time_val)
       break
```
The for-loop could be aborted as soon as a non-zero value is encountered. So, I am not sure what will be more efficient. In any case, we can drop has_units and replace the following loop over `datetime_delta`s with a look-up in the `unique_units_or_whatever` dict...

echoix replied:
> Looks like this basically works but v.add returns None.
> 
> 
> 
> So, this could do the trick:
> 
> ```
> unique_units_or_whatever = {
>     'hour': set(),
>     'minute': set(),
>     'second': set(),
>     }
> 
> 
> 
> map_datetime_delta = {"hour": 2, "second": 2, "minute": 0}
> 
> 
> 
> # Then we can either union two sets and replace the dict from comprehension
> 
> unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items() }
> 
> 
> 
> # or loop over ordered keys and update in place:
> 
> for time_key in ["second", "minute", "hour"]:
> 
>     time_val = map_datetime_delta[time_key]
> 
>     if time_val > 0:
> 
>        unique_units_or_whatever[time_key].add(time_val)
> 
>        break
> 
> ```
> 
> The for-loop could be aborted as soon as a non-zero value is encountered. So, I am not sure what will be more efficient. In any case, we can drop has_units and replace the following loop over `datetime_delta`s with a look-up in the `unique_units_or_whatever` dict...


In `unique_units_or_whatever` it was supposed to be all the possible units, I only written 3 to see the pattern. Just making sure we understood the same thing. 

```python
 unique_units_or_whatever = {
     'hour': set(),
     'minute': set(),
     'second': set(),     
     'max_days': set(),
     ...
     'years': set()
  }
```


Nothing prevents having the condition inside the dict comprehension, like:
```python

unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items()if map_datetime_delta[k] > 0}

```

But given how many times that loop would be hit, are we able to handle having 0 in the set, and still being able to do the gcd? (If gcd doesn't work, the. We could remove the 0 value before, with a set operation).
If so, we could initialize the sets with `set(0)`, or `{0,}` and probably that v.add wouldn't be None. I would try to use word operators here rather than only characters, since it's easier to read back, and makes sure the correct operation is done if there's a problem with the types (like what if v is initialized with a bool).
The operation could be something in https://docs.python.org/3/library/stdtypes.html#frozenset.union that works. 
I didn't find back the source document that explained it better than 
> Note, the non-operator versions of the update(), intersection_update(), difference_update(), and symmetric_difference_update() methods will accept any iterable as an argument.



And for `v | {map_datetime_delta[k]}`, if it was possible to not create another set litteral for it to work it might be better.

echoix replied:
And I forgot to explain why I wanted to go with a dict/list/set comprehension: they usually have a little edge performance wise, especially on startup performance. It's been a long time that it has been better. I wonder if the differences with other loops are still as big with Python 3.10+, but it's a good trick to know.

In Python 2 to 3 era, I remember working on a scientific code base (with a pyqt gui), that heavily relied on "identity"-like dict comprehension to copy dictionaries to work around the cases where a shallow copy/deep copy would have been used. There was quite some uses for it!

ninsbl replied:
Initial comparisons of the above dict-comprehension vs. for-loop indicate that the for-loop is faster esp. if the time delta contains seconds (and looping stops early)....

The if-clause in the dict-comprehension does not really work as keys disappear...

But I will work more on this issue later...

echoix replied:
That would mean that instead of assigning to the same dictionary we would need to use the dict.update() method. That would only override the values from keys that are in the new dict, but it won't remove the older ones. So it would change probably one or two keys per map, depending on the non 0 ones.

You can go with a for loop that works with a note that it could be optimized in another round. It's getting a bit of a long review, and as long as it works better in the first place (and it is readable)

echoix reviewed (commented):
I'm not sure that I passed through all the granularity file, but it's getting late. And there was some easy and harder fixes.

ninsbl commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,143 @@ def compute_absolute_time_granularity(maps):
+        start, end = get_time_tuple(stds_map)
+        # start time is required in TGIS and expected to be present
+        if end:
+            map_datetime_delta = compute_datetime_delta(start, end)
```
It is fully possible to have a temporal granularity of "1 minute 30 seconds" or or "3 days 1 hour 1 second" or "1 year 6 months"...

I agree that the concepts of relative vs. absolute time could be better documented.
I have not used that type too much myself...

The [wiki page](https://grasswiki.osgeo.org/wiki/Temporal_data_processing) says:

>  the relative temporal type refers to data without fixed time stamps (e.g., sequential maps used to calculate multi-decadal averages)

echoix replied:
I agree that a granularity could be 1 min 30 sec. But shouldn't it be 1 min 30 sec for


```python

>>> start = datetime(2001, 1, 1, 00,00,14)
         >>> end = datetime(2001, 1, 1, 00,00,44)
         >>> compute_datetime_delta(start, end)
         {'hour': 0, 'month': 0, 'second': 30, 'max_days': 0, 'year': 0, 'day': 0, 'minute': 0}

         >>> start = datetime(2001, 1, 1, 00,00,44)
         >>> end = datetime(2001, 1, 1, 00,01,14)
         >>> compute_datetime_delta(start, end)
         {'hour': 0, 'month': 0, 'second': 30, 'max_days': 0, 'year': 0, 'day': 0, 'minute': 0}

 
         >>> start = datetime(2001, 1, 1, 00,00,44)
         >>> end = datetime(2001, 1, 1, 00,02,14)
         >>> compute_datetime_delta(start, end)
         {'hour': 0, 'month': 0, 'second': 30, 'max_days': 0, 'year': 0, 'day': 0, 'minute': 1}

 OR

         {'hour': 0, 'month': 0, 'second': 90, 'max_days': 0, 'year': 0, 'day': 0, 'minute': 1}

OR

         {'hour': 0, 'month': 0, 'second': 90, 'max_days': 0, 'year': 0, 'day': 0, 'minute': 0}

```

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -111,6 +112,81 @@ def check_granularity_string(granularity, temporal_type):
+
+def _is_after(start, start1, end1):
+    """Helper function that checks if start timestamp is
+    temporaly after the start1 and end1, where start1 and end1
```
```suggestion
    temporally after the start1 and end1, where start1 and end1
```

ninsbl replied:
Fixed in [04d4778](https://github.com/OSGeo/grass/pull/3350/commits/04d4778d32e34a4c3842187fd22bb7fbc4a8d34e)

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,83 @@ def compute_absolute_time_granularity(maps):
-    datetime_delta = []
+    # Create a granularity dict with time units of increasing length
+    # that covers all possible keys in the result of compute_datetime_delta
+    # The order of the keys is important so that loops over the dictionalry
```
```suggestion
    # The order of the keys is important so that loops over the dictionary
```

ninsbl replied:
Fixed in [04d4778](https://github.com/OSGeo/grass/pull/3350/commits/04d4778d32e34a4c3842187fd22bb7fbc4a8d34e)

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,83 @@ def compute_absolute_time_granularity(maps):
-            return "%i year" % granularity
-        else:
-            return "%i years" % granularity
+    if granularity is None or time_unit is None:
```
In an IDE with pyright, it says time_unit is possibly unbound. And effectively it is always used as a loop variable, never really defined.

ninsbl replied:
Right, changed in in [04d4778](https://github.com/OSGeo/grass/pull/3350/commits/04d4778d32e34a4c3842187fd22bb7fbc4a8d34e)

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,83 @@ def compute_absolute_time_granularity(maps):
+        "minute": set(),
+        "hour": set(),
+        "day": set(),
+        "max_days": set(),
```
What does max days represent? some tests fail with max days as a unit. Before, it wasn't a possible option.

If the goal is to make them synonyms for computign the gcd, it could be combined into days (and other conversions needed) before calling gcd at the end.

ninsbl replied:
`max_days` was tricky..,

I don`t think I fully understand the purpose of that "unit", but it is now handled properly, I guess...

echoix commented on python/grass/temporal/temporal_granularity.py:
```diff
@@ -344,164 +423,83 @@ def compute_absolute_time_granularity(maps):
-    ulist = list(set(dlist))
-
-    if len(ulist) == 0:
+    dlist = set()
+    for time_unit, granularity_set in granularity_units.items():
+        # The smallest granularity unit is used so as soon as a non-zero
+        # value / set is encountered, the loop can be aborted
+        if granularity_set:
+            dlist.update(granularity_set)
+            break
+
```
Before there was a handling of converting bigger "units" into the smaller one before gcd.
How does it create an equivalent effect now?
(And I'm still not convinced that `compute_datetime_delta` does what I think it should do, so don't let me bring you to a bad solution because of my comprehension).

ninsbl replied:
Yeah, seems I overshot the mark when simplified too much... Fixed in 04d4778.

ninsbl commented:
@landam or @neteler , do you see any issues with the change here:
https://github.com/OSGeo/grass/pull/3350/files#diff-b82f893a8015e0a5f7eb01a45216a2292d295d9c1dab66e2176cbe48610eca03

Any objections?

Tests don`t fail, however, older TGIS DBs are not tested...
From reading the code I assume a lack backwards compatibility with older TGIS databases is discovered already during initialisation: https://github.com/OSGeo/grass/blob/04d4778d32e34a4c3842187fd22bb7fbc4a8d34e/python/grass/temporal/core.py#L751
No?

echoix reviewed (approved):
If the code does the correct thing, I'm ok with the how it is made. It is a good improvement.

ninsbl commented:
> If the code does the correct thing, I'm ok with the how it is made. It is a good improvement.

Great! Thanks for all your effort @echoix ! Very much appreciated! I will merge a bit later today...

echoix commented:
Is there some topic that you are comfortable reviewing? Is there any of my open PRs that you can take a look at?

ninsbl commented:
> Is there some topic that you are comfortable reviewing? Is there any of my open PRs that you can take a look at?

Sure. Will have a look at it a bit later today...

wenzeslaus commented:
> At some places using context managers (that can close connections) might help a bit with the dbif always getting created and closed inside a same function

Yes, context managers are quite helpful and applicable in our code. (And did not exist in the past.)
